File: Semester3/Advanced Programming Methods/Labs/lab2/src/task_pck/QuickSort.java
package task_pck;

import java.util.ArrayList;
import java.util.Collections;

public class QuickSort extends AbstractSorter {
    @Override
    public ArrayList<Integer> sort(ArrayList<Integer> arr) {
        arr = quick_sort(0, arr.size() - 1, arr);
        return arr;
    }

    private ArrayList<Integer> quick_sort(int left, int right, ArrayList<Integer> arr) {

        if(right <= left){
            return arr;
        }

        int mid = (left + right) / 2;
        int pivot = arr.get(mid);
        int i = left, j = right;

        while(i < j){
            while(arr.get(i).compareTo(pivot) < 0){
                i ++;
            }
            while(arr.get(j).compareTo(pivot) > 0){
                j --;
            }

            if(i < j){
                Collections.swap(arr, i, j);
            }
        }

        quick_sort(left, j - 1, arr);
        quick_sort(j + 1, right, arr);

        return arr;
    }
}

File: BachelorExamPrep/OOP/Execices/quickSort.cpp
//
// Created by geo on 17.06.2020.
//

#include "quickSort.h"


int partition(int *arr, int left, int right) {
    int pivot = arr[right];
    int i = left - 1;
    for (int j = left; j < right; j++) {
        if (arr[j] < pivot) {
            i++;
            int tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }
    }

    int tmp = arr[i + 1];
    arr[i + 1] = arr[right];
    arr[right] = tmp;

    return i + 1;
}

void quickSort(int left, int right, int *arr) {
    if (left < right) {
        int pIndex = partition(arr, left, right);

        quickSort(left, pIndex - 1, arr);
        quickSort(pIndex + 1, right, arr);
    }

    //partitionare
}